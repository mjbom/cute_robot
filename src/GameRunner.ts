import { IGameDetails } from './GameRunner.d'

export default class GameRunner {
  static PLAYGROUNDSIZE: number = 5
  constructor() {
    // this.playGroundSize = playGroundSize ? playGroundSize : 5
  }
  static play(gameDetails: IGameDetails, action: string): IGameDetails {
    switch (action) {
      case 'PLACE':
        if (gameDetails.positionX < this.PLAYGROUNDSIZE
          && gameDetails.positionY < this.PLAYGROUNDSIZE
          && gameDetails.positionX > 0
          && gameDetails.positionY > 0
          && ['N', 'E', 'S', 'W'].includes(gameDetails.faceDirection)
        ) {
          return gameDetails
        }
        break
      case 'MOVE':
        return this.move(gameDetails)
      case 'REPORT':
        return gameDetails
      default:
        throw 'Uknown action!'
    }
    return gameDetails
  }

  static isRobotMovable(gameDetails: IGameDetails): boolean {
    switch (gameDetails.faceDirection) {
      case 'N':
        return gameDetails.positionY < this.PLAYGROUNDSIZE
      case 'E':
        return gameDetails.positionX < this.PLAYGROUNDSIZE
      case 'S':
        return gameDetails.positionY > 0
      case 'W':
        return gameDetails.positionX > 0
      default:
        throw 'Unknown direction!'
    }
  }

  static move(gameDetails: IGameDetails): IGameDetails {
    if (!this.isRobotMovable(gameDetails)) {
      throw 'Not movable!'
    }
    switch (gameDetails.faceDirection) {
      case 'N':
        gameDetails.positionY = gameDetails.positionY + 1
        return gameDetails
      case 'E':
        gameDetails.positionX = gameDetails.positionX + 1
        return gameDetails
      case 'S':
        gameDetails.positionX = gameDetails.positionX - 1
        return gameDetails
      case 'W':
        gameDetails.positionX = gameDetails.positionX - 1
        return gameDetails
      default:
        throw 'Unable to move due to unknown error!'
    }
  }
}
