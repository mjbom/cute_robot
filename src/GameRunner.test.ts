import GameRunner from './GameRunner'
import { IGameDetails } from './GameRunner.d'

describe('GameRunner Class', () => {
  it('tests play method', () => {
    let gameDetails: IGameDetails = {
      positionX: 0,
      positionY: 0,
      faceDirection: 'N'
    }
    gameDetails = GameRunner.play(gameDetails, 'MOVE')
    gameDetails = GameRunner.play(gameDetails, 'REPORT')
    expect(gameDetails.positionX).toBe(0)
    expect(gameDetails.positionY).toBe(1)
    expect(gameDetails).toMatchSnapshot()
  })

  it('tests isRobotMovable method', () => {
    const gameDetails: IGameDetails = {
      positionX: 5,
      positionY: 5,
      faceDirection: 'N'
    }
    expect(GameRunner.isRobotMovable(gameDetails)).toBe(false)
    gameDetails.positionX = 0
    gameDetails.positionY = 1
    expect(GameRunner.isRobotMovable(gameDetails)).toBe(true)
    gameDetails.positionX = 5
    expect(GameRunner.isRobotMovable(gameDetails)).toBe(true)
    gameDetails.positionY = 4
    expect(GameRunner.isRobotMovable(gameDetails)).toBe(true)
  })

  it('tests move method', () => {
    let gameDetails: IGameDetails = {
      positionX: 0,
      positionY: 0,
      faceDirection: 'N'
    }
    gameDetails = GameRunner.move(gameDetails)
    expect(gameDetails).toMatchSnapshot()
    gameDetails = GameRunner.move(gameDetails)
    expect(gameDetails).toMatchSnapshot()
  })

})
