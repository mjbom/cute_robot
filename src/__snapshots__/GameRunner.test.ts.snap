// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameRunner Class tests move method 1`] = `
Object {
  "faceDirection": "N",
  "positionX": 0,
  "positionY": 1,
}
`;

exports[`GameRunner Class tests move method 2`] = `
Object {
  "faceDirection": "N",
  "positionX": 0,
  "positionY": 2,
}
`;

exports[`GameRunner Class tests play method 1`] = `
Object {
  "faceDirection": "N",
  "positionX": 0,
  "positionY": 1,
}
`;
